### DISCUSSION TITLE
Keeping the cursor in the current position

### DISCUSSION THREAD

**Discussion by navlalli** [michaelb/sniprun#200] [Q&A] [answered]
Firstly, fantastic plugin @michaelb ! 

I have defined a keymap in a *.lua* file which allows me to run the whole file by pressing F5:
```
vim.keymap.set('n', '<F5>', ":%SnipRun <CR>", {})
```

I've been trying to set this keymap up so that the cursor remains at the current position when I use this shortcut (right now it jumps to the top of the file and `<Ctrl-o>` doesn't always take me back to the same position). I've been trying to use `let b:caret=winsaveview()` and `call winrestview(b:caret)`, as written in the [documentation](https://michaelb.github.io/sniprun/sources/README.html#:~:text=let%20b%3Acaret%3Dwinsaveview()%20and%20call%20winrestview(b%3Acaret)), but with no success so far.

Has anyone got an example of how to achieve this in a *.lua* file?

**Comment by michaelb** [comment, marked as answer]
Something like

` :lua vim.keymap.set('n', '<F5>', ":let b:caret=winsaveview() <CR> | :%SnipRun <CR>| :call winrestview(b:caret) <CR>", {}) ` should do the trick (it does for me), though there may be better ways to express that in vimscript and I'm no expert

**Reply by navlalli** [reply]
Works perfectly, thank you!! May be useful to add something like that below [this](https://michaelb.github.io/sniprun/sources/README.html#:~:text=For%20interpreted%20languages,most%20of%20them) part in the documentation.

**Reply by rinpatch** [reply]
@michaelb Hi, I am trying to adapt your code for running a snipped from start of file to current line. Here is what I got:
```lua
vim.keymap.set('n', '<leader>fb', ":1,. SnipRun <CR>", {})
vim.keymap.set('n', '<leader>fv', ":let b:caret=winsaveview() <CR> | :1,. SnipRun <CR> | :call winrestview(b:caret) <CR>", {})
vim.keymap.set('n', '<leader>fc', ":let b:caret=winsaveview() <CR> | :1,. SnipRun <CR> | :sleep 1 | :call winrestview(b:caret) <CR>", {})
````

With this example python code:
```python
test = "test"

def fun(test):
    return test 

print(fun(test))
```

And cursor positioned at `4,1`, The function results are:
- `<leader>fb`: Outputs `(no output)` and puts the cursor on top.
- `<leader>fv`: Outputs `RuntimeError: SyntaxError: 'return' outside function`, cursor stays in place.
- `<leader>fc`: Outputs `(no output)`, puts the cursor on top for one second, then moves it back to where it was.

I don't get why `<leader>fv` doesn't work correctly, but works if I add a sleep. Please help me figure this out.

**Reply by navlalli** [reply]
Hey @rinpatch 
I believe this is more what you are looking for where this will run from lines 1 to 5 and should keep your cursor in the same position:
```
vim.keymap.set('n', '<leader>fv ', ":let b:caret=winsaveview() <CR> | :lua require'sniprun.api'.run_range(1, 5) <CR>| :call winrestview(b:caret) <CR>", {})
```
However, I wasn't able to figure out how to insert the current line number into `run_range(1, 5)` but hopefully there should be a way.

**Reply by rinpatch** [reply]
@navlalli I added the current line number like this:
```lua
vim.keymap.set('n', '<leader>fv', ":let b:caret=winsaveview() <CR> | :lua require'sniprun.api'.run_range(1, vim.api.nvim_win_get_cursor(0)[1]) <CR> | :call winrestview(b:caret) <CR>", {})
```
Seems to work perfectly so far. Thank you.

**Reply by michaelb** [reply]
Well done both of you 😂, I couldn't have suggested better answers.

This winsaveview() shenanigan is a bit ugly though, so I'll look into it someday
